crates/stages/src/stages/execution.rs

```diff
@@ -421,17 +427,39 @@ impl<EF: ExecutorFactory, DB: Database> Stage<DB> for ExecutionStage<EF> {
         let mut stage_checkpoint = input.checkpoint.execution_stage_checkpoint();
 
         // Unwind all receipts for transactions in the block range
-        let mut cursor = tx.cursor_write::<tables::Receipts>()?;
-        let mut reverse_walker = cursor.walk_back(None)?;
-
-        while let Some(Ok((tx_number, receipt))) = reverse_walker.next() {
-            if tx_number < first_tx_num {
-                break
+        match try_prepare_snapshotter(&self.prune_modes, provider, *range.start())? {
+            Some(_) => {
+                // `try_prepare_snapshotter` already prunes receipts if it detects that there are
+                // more elements in the static files than expected by block `range.start()`,
+                // effectively unwinding it.
+
+                // Update the checkpoint.
+                if let Some(stage_checkpoint) = stage_checkpoint.as_mut() {
+                    for block_number in range {
+                        stage_checkpoint.progress.processed -= provider
+                            .block_by_number(block_number)?
+                            .ok_or_else(|| ProviderError::BlockNotFound(block_number.into()))?
+                            .gas_used;
+                    }
+                }
             }
-            reverse_walker.delete_current()?;
+            None => {
+                // `try_prepare_snapshotter` returning None means that the node has some kind of
+                // filtering/pruning enabled. In this case static files are not used, and falls back
+                // to database.
+                let mut cursor = tx.cursor_write::<tables::Receipts>()?;
+                let mut reverse_walker = cursor.walk_back(None)?;
+
+                while let Some(Ok((tx_number, receipt))) = reverse_walker.next() {
+                    if tx_number < first_tx_num {
+                        break
+                    }
+                    reverse_walker.delete_current()?;
 
-            if let Some(stage_checkpoint) = stage_checkpoint.as_mut() {
-                stage_checkpoint.progress.processed -= receipt.cumulative_gas_used;
+                    if let Some(stage_checkpoint) = stage_checkpoint.as_mut() {
+                        stage_checkpoint.progress.processed -= receipt.cumulative_gas_used;
```

Comment:
why does this `-=`?

Reply:
because this is part of the unwind and we need to set the total gas progress back .

it already existed beforehand, this is my understanding